python/ql/src/experimental/dataflow/TypeTracker.qll

```diff
@@ -0,0 +1,283 @@
+    or
+    exists(string prop | this = StoreStep(prop) | result = "store " + prop)
+    or
+    exists(string prop | this = LoadStep(prop) | result = "load " + prop)
+  }
+}
+
+module StepSummary {
+  cached
+  predicate step(Node pred, Node succ, StepSummary summary) {
+    exists(Node mid | EssaFlow::essaFlowStep*(pred, mid) and smallstep(mid, succ, summary))
+  }
+
+  predicate smallstep(Node pred, Node succ, StepSummary summary) {
+    EssaFlow::essaFlowStep(pred, succ) and
+    summary = LevelStep()
+    or
+    callStep(pred, succ) and summary = CallStep()
+    or
+    returnStep(pred, succ) and
+    summary = ReturnStep()
+    or
+    exists(string attr |
+      basicStoreStep(pred, succ, attr) and
+      summary = StoreStep(attr)
+      or
+      basicLoadStep(pred, succ, attr) and summary = LoadStep(attr)
+    )
+  }
+}
+
+/** Holds if `pred` steps to `succ` by being passed as a parameter in a call. */
+predicate callStep(ArgumentNode pred, ParameterNode succ) {
+  // TODO: Support special methods?
+  exists(DataFlowCall call, int i |
+    pred.argumentOf(call, i) and succ.isParameterOf(call.getCallable(), i)
+  )
+}
+
+/** Holds if `pred` steps to `succ` by being returned from a call. */
+predicate returnStep(ReturnNode pred, Node succ) {
+  exists(DataFlowCall call |
+    pred.getEnclosingCallable() = call.getCallable() and succ = TCfgNode(call.getNode())
+  )
+}
+
+/**
+ * Holds if `pred` is being written to the `attr` attribute of the object in `succ`.
+ *
+ * Note that the choice of `succ` does not have to make sense "chronologically".
+ * All we care about is whether the `attr` attribute of `succ` can have a specific type,
+ * and the assumption is that if a specific type appears here, then any access of that
+ * particular attribute can yield something of that particular type.
+ *
+ * Thus, in an example such as
+ *
+ * ```python
+ * def foo(y):
+ *    x = Foo()
+ *    bar(x)
+ *    x.attr = y
+ *    baz(x)
+ *
+ * def bar(x):
+ *    z = x.attr
+ * ```
+ * for the attribute write `x.attr = y`, we will have `attr` being the literal string `"attr"`,
+ * `pred` will be `y`, and `succ` will be the object `Foo()` created on the first line of the
+ * function. This means we will track the fact that `x.attr` can have the type of `y` into the
+ * assignment to `z` inside `bar`, even though this attribute write happens _after_ `bar` is called.
+ */
+predicate basicStoreStep(Node pred, Node succ, string attr) {
+  exists(AttributeAssignment a, Node var |
+    a.getName() = attr and
+    EssaFlow::essaFlowStep*(succ, var) and
+    var.asVar() = a.getInput() and
+    pred.asCfgNode() = a.getValue()
+  )
+}
+
+/**
+ * Holds if `succ` is the result of accessing the `attr` attribute of `pred`.
+ */
+predicate basicLoadStep(Node pred, Node succ, string attr) {
+  exists(AttrNode s | succ.asCfgNode() = s and s.getObject(attr) = pred.asCfgNode())
+}
+
+/**
+ * A utility class that is equivalent to `boolean` but does not require type joining.
```

Comment:
This comes straight from the JavaScript libraries, so I'm not 100% what the intent is, but I think the issue is that if you write
```ql
private newtype TTypeTracker = MkTypeTracker(boolean hasCall, OptionalAttributeName prop)
```
(note lowercase `boolean`) the compiler will complain about `hasCall` not being sufficiently bound. This is because `boolean` is treated the same way as, say, `int` or `string` for the purposes of checking binding, even though `boolean` only has two possible values. By encapsulating it in a utility class, we avoid having to specify `hasCall = true or hasCall = false` in the above definition. (Which, in isolation, is probably not a large gain, but I imagine it could be useful if we start using booleans elsewhere.)

Reply:
right, I see. Thanks. If you could add some of that nice explanation into the comment in the code that could be good! :muscle: